src/leiningen/ring/uberwar.clj

```diff
@@ -11,27 +14,45 @@
 (defn get-classpath [project]
   (if-let [get-cp (resolve 'leiningen.core.classpath/get-classpath)]
     (get-cp project)
-    (->> (:library-path project) io/file .listFiles (map str))))
-
-(defn contains-entry? [^java.io.File file ^String entry]
-  (with-open [jar-file (JarFile. file)]
-    (some (partial = entry)
-          (map #(.getName ^JarEntry %)
-               (enumeration-seq (.entries jar-file))))))
+    (->> (:library-path project) (io/file) .listFiles (map str))))
 
 (defn jar-dependencies [project]
   (for [pathname (get-classpath project)
-        :let [file (io/file pathname)
-              fname (.getName file)]
-        :when (and (.endsWith fname ".jar")
-                   ;; Servlet container will have it's own servlet-api impl
-                   (not (contains-entry? file "javax/servlet/Servlet.class")))]
-    file))
-
-(defn jar-entries [war project]
+        :when (str/ends-with? pathname ".jar")]
+    (io/file pathname)))
+
+(def ^:private javax-servlet?  #{"javax/servlet/Servlet.class"})
+(def ^:private pom-properties? #(str/ends-with? % "pom.properties"))
+
+(defn- war-path-for-jar
+  [^java.io.File jar]
+  (with-open [jar-file (JarFile. jar)]
+    (let [javax&pom (->> (.entries jar-file)
+                         enumeration-seq
+                         (map (fn [^JarEntry je] (.getName je)))
+                         (filter (some-fn javax-servlet? pom-properties?)))] ;; 2 elements maximum
+      ;; Servlet container will have it's own servlet-api impl
+      (when-not (some javax-servlet? javax&pom)
+        ;; we've confirmed that one of the two possible files is missing,
+        ;; so if we find one (via `first`), it will be the pom.properties
+        (let [pom-properties-full-path (some->> (first javax&pom)
+                                                (str "jar:file:" (.getAbsolutePath jar) "!/"))
+              {:strs [groupId artifactId version]} (when pom-properties-full-path
+                                                     (with-open [in (.openStream (URL. pom-properties-full-path))]
+                                                       (->> (doto (Properties.) (.load in))
+                                                            (into {}))))]
+          (->> (if (and groupId artifactId version)
+                 [groupId \- artifactId \- version ".jar"] ;; found pom.properties - reconstruct the jar name to avoid collisions (issue #216)
+                 [(.getName jar)])                         ;; fallback to using the original jar name (best-effort approach)
+               (apply str "WEB-INF/lib/")))))))
+
+(defn jar-entries
```

Comment:
Is this intended to be part of the public API?

Reply:
no, it can be private